lib/pronto/formatter/git_formatter_base.rb

```diff
@@ -0,0 +1,88 @@
+module Pronto
+  module Formatter
+    class GitFormatterBase
+      def format(messages, repo, patches)
+        client    = client_module.new(repo)
+        head      = repo.head_commit_sha
+        existing  = existing_comments(client, head)
+        comments  = new_comments(messages, patches, head)
+        additions = remove_duplicate_comments(existing, comments)
+        submit_comments(client, additions)
+
+        "#{additions.count} Pronto messages posted to #{pretty_name}"
+      end
+
+      def client_module
+        raise NotImplementedError
+      end
+
+      def pretty_name
+        raise NotImplementedError
+      end
+
+      protected
+
+      def existing_comments(client, sha)
+        raise NotImplementedError
+      end
+
+      def line_number(message)
+        raise NotImplementedError
+      end
+
+      def submit_comments(client, comments)
+        raise NotImplementedError
+      end
+
+      private
+
+      def config
+        @config ||= Config.new
+      end
+
+      def consolidate_comments(comments)
+        result = comments.first
+        if comments.length > 1
+          joined_body = comments
+                        .map { |comment| "- #{comment.body}" }
+                        .join("\n")
+
+          result = client_module::Comment.new(result.sha, joined_body, result.path, result.position)
+        end
+        result
+      end
+
+      def new_comment(message, patches, sha)
+        body = message.msg
+        path = message.path
+        lineno = line_number(message)
+
+        client_module::Comment.new(sha, body, path, lineno)
+      end
+
+      def new_comments(messages, patches, sha)
+        messages
+          .uniq
+          .map { |message| new_comment(message, patches, sha) }
+          .group_by { |comment| [comment.path, comment.position] }
```

Comment:
This `group_by` is repeated multiple times. Maybe there's a way to reduce duplication?